File: modules/builder/src/builders/xviz-builder.js
// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Note: XVIZ data structures use snake_case
/* eslint-disable camelcase */
import XVIZPoseBuilder from './xviz-pose-builder';
import XVIZLinkBuilder from './xviz-link-builder';
import XVIZPrimitiveBuilder from './xviz-primitive-builder';
import XVIZFutureInstanceBuilder from './xviz-future-instance-builder';
import XVIZUIPrimitiveBuilder from './xviz-ui-primitive-builder';
import XVIZTimeSeriesBuilder from './xviz-time-series-builder';
import XVIZValidator from './xviz-validator';
import XVIZVariableBuilder from './xviz-variable-builder';
import {PRIMARY_POSE_STREAM} from './constant';

/* global console */
/* eslint-disable no-console */
const defaultValidateWarn = console.warn;
const defaultValidateError = console.error;
/* eslint-enable no-console */

// TODO: Builder could validate against stream metadata!
export default class XVIZBuilder {
  constructor({
    metadata = {},
    disableStreams = [],
    validateWarn = defaultValidateWarn,
    validateError = defaultValidateError
  } = {}) {
    this._validator = new XVIZValidator({
      validateWarn,
      validateError
    });

    this.metadata = metadata;
    this.disableStreams = disableStreams;

    this.updateType = 'SNAPSHOT';

    // Current streamBuilder
    this._streamBuilder = null;

    // Construct different builders
    this._poseBuilder = new XVIZPoseBuilder({
      metadata: this.metadata,
      validator: this._validator
    });
    this._variablesBuilder = new XVIZVariableBuilder({
      metadata: this.metadata,
      validator: this._validator
    });
    this._primitivesBuilder = new XVIZPrimitiveBuilder({
      metadata: this.metadata,
      validator: this._validator
    });
    this._futureInstanceBuilder = new XVIZFutureInstanceBuilder({
      metadata: this.metadata,
      validator: this._validator
    });
    this._uiPrimitivesBuilder = new XVIZUIPrimitiveBuilder({
      metadata: this.metadata,
      validator: this._validator
    });
    this._timeSeriesBuilder = new XVIZTimeSeriesBuilder({
      metadata: this.metadata,
      validator: this._validator
    });
    this._linkBuilder = new XVIZLinkBuilder({
      metadata: this.metadata,
      validator: this._validator
    });
  }

  persistent() {
    this.updateType = 'PERSISTENT';
  }

  pose(streamId = PRIMARY_POSE_STREAM) {
    this._streamBuilder = this._poseBuilder.stream(streamId);
    return this._streamBuilder;
  }

  link(parent, child) {
    this._streamBuilder = this._linkBuilder.stream(child).parent(parent);
    return this._streamBuilder;
  }

  variable(streamId) {
    this._streamBuilder = this._variablesBuilder.stream(streamId);
    return this._streamBuilder;
  }

  primitive(streamId) {
    this._streamBuilder = this._primitivesBuilder.stream(streamId);
    return this._streamBuilder;
  }

  futureInstance(streamId, timestamp) {
    this._streamBuilder = this._futureInstanceBuilder.stream(streamId);
    this._streamBuilder._timestamp(timestamp);
    return this._streamBuilder;
  }

  uiPrimitive(streamId) {
    this._streamBuilder = this._uiPrimitivesBuilder.stream(streamId);
    return this._streamBuilder;
  }

  timeSeries(streamId) {
    this._streamBuilder = this._timeSeriesBuilder.stream(streamId);
    return this._streamBuilder;
  }

  /*
  message data:
  {
    update_type: 'SNAPSHOT',
    updates: [{
      timestamp,
      poses: {'/vehicle-pose': {}, ...},
      primitives: {},
      variables: {},
      future_instances: {}
    }]
  }
   */
  getMessage() {
    const {poses} = this._poseBuilder.getData();

    if (!poses || !poses[PRIMARY_POSE_STREAM]) {
      this._validator.error(`Every message requires a ${PRIMARY_POSE_STREAM} stream`);
    }

    const primitives = this._primitivesBuilder.getData();
    const futures = this._futureInstanceBuilder.getData();
    const variables = this._variablesBuilder.getData();
    const timeSeries = this._timeSeriesBuilder.getData();
    const uiPrimitives = this._uiPrimitivesBuilder.getData();
    const links = this._linkBuilder.getData();

    const data = {
      timestamp: poses[PRIMARY_POSE_STREAM].timestamp,
      poses
    };

    if (primitives) {
      data.primitives = primitives;
    }
    if (futures) {
      data.future_instances = futures;
    }
    if (variables) {
      data.variables = variables;
    }
    if (timeSeries) {
      data.time_series = timeSeries;
    }
    if (uiPrimitives) {
      data.ui_primitives = uiPrimitives;
    }
    if (links) {
      data.links = links;
    }

    const message = {
      update_type: this.updateType,
      updates: [data]
    };

    return message;
  }

  // DEPRECATED: change to using getMessage()
  getFrame() {
    return this.getMessage();
  }

  _reset() {
    this._streamBuilder = null;
  }
}

File: python/xviz_avs/builder/xviz_builder.py
import logging
from easydict import EasyDict as edict

from xviz_avs.message import XVIZFrame, XVIZMessage

from xviz_avs.builder.link import XVIZLinkBuilder
from xviz_avs.builder.future_instance import XVIZFutureInstanceBuilder
from xviz_avs.builder.pose import XVIZPoseBuilder
from xviz_avs.builder.primitive import XVIZPrimitiveBuilder
from xviz_avs.builder.variable import XVIZVariableBuilder
from xviz_avs.builder.ui_primitive import XVIZUIPrimitiveBuilder
from xviz_avs.builder.time_series import XVIZTimeSeriesBuilder

from xviz_avs.v2.core_pb2 import StreamSet
from xviz_avs.v2.session_pb2 import StateUpdate
from google.protobuf.json_format import MessageToDict

PRIMARY_POSE_STREAM = '/vehicle_pose'

class XVIZBuilder:
    def __init__(self, metadata=None, disable_streams=None,
                 logger=logging.getLogger("xviz")):
        self._logger = logger
        self._metadata = metadata
        self._disable_streams = disable_streams or []
        self._stream_builder = None
        self._update_type = StateUpdate.UpdateType.INCREMENTAL

        self._links_builder = XVIZLinkBuilder(self._metadata, self._logger)
        self._pose_builder = XVIZPoseBuilder(self._metadata, self._logger)
        self._variables_builder = XVIZVariableBuilder(self._metadata, self._logger)
        self._primitives_builder = XVIZPrimitiveBuilder(self._metadata, self._logger)
        self._future_instance_builder = XVIZFutureInstanceBuilder(self._metadata, self._logger)
        self._ui_primitives_builder = XVIZUIPrimitiveBuilder(self._metadata, self._logger)
        self._time_series_builder = XVIZTimeSeriesBuilder(self._metadata, self._logger)

    def pose(self, stream_id=PRIMARY_POSE_STREAM):
        self._stream_builder = self._pose_builder.stream(stream_id)
        return self._stream_builder

    def variable(self, stream_id):
        self._stream_builder = self._variables_builder.stream(stream_id)
        return self._stream_builder

    def primitive(self, stream_id):
        self._stream_builder = self._primitives_builder.stream(stream_id)
        return self._stream_builder

    def future_instance(self, stream_id, timestamp):
        self._stream_builder = self._future_instance_builder.stream(stream_id)
        self._stream_builder.timestamp(timestamp)
        return self._stream_builder

    def ui_primitives(self, stream_id):
        self._stream_builder = self._ui_primitives_builder.stream(stream_id)
        return self._stream_builder

    def time_series(self, stream_id):
        self._stream_builder = self._time_series_builder.stream(stream_id)
        return self._stream_builder

    def link(self, parent, child):
        self._stream_builder = self._links_builder.stream(child).parent(parent)
        return self._stream_builder

    def _reset(self):
        self._stream_builder = None

    def get_data(self):
        poses = self._pose_builder.get_data()
        if (not poses) or (PRIMARY_POSE_STREAM not in poses):
            self._logger.error('Every message requires a %s stream', PRIMARY_POSE_STREAM)

        data = XVIZFrame(StreamSet(
            timestamp=poses[PRIMARY_POSE_STREAM].timestamp, # FIXME: does timestamp have to be the same with pose?
            poses=poses,
            primitives=self._primitives_builder.get_data(),
            future_instances=self._future_instance_builder.get_data(),
            variables=self._variables_builder.get_data(),
            time_series=self._time_series_builder.get_data(),
            ui_primitives=self._ui_primitives_builder.get_data(),
            links=self._links_builder.get_data()
        ))

        return data

    def get_message(self):
        message = XVIZMessage(StateUpdate(
            update_type=self._update_type,
            updates=[self.get_data().data]
        ))
        return message
